File: app/src/main/java/com/example/whattodo/HomeFragment.java
package com.example.whattodo;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.location.LocationSettingsStatusCodes;
import com.google.android.gms.location.SettingsClient;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.SetOptions;
import com.google.firebase.messaging.FirebaseMessaging;

import java.io.IOException;
import java.text.DateFormat;
import java.text.Normalizer;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class HomeFragment extends Fragment implements View.OnClickListener{

    private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 10000;
    private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS =
            UPDATE_INTERVAL_IN_MILLISECONDS / 2;
    private static final String TAG = MainActivity.class.getSimpleName();
    private static final int REQUEST_CHECK_SETTINGS = 0x1;
    private static final int LOCATION_PERMISSION_REQUEST_CODE = 34;
    private static final String FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION;
    private static final String COURSE_LOCATION = Manifest.permission.ACCESS_COARSE_LOCATION;
    private final static String KEY_REQUESTING_LOCATION_UPDATES = "requesting-location-updates";
    private final static String KEY_LOCATION = "location";
    private final static String KEY_LAST_UPDATED_TIME_STRING = "last-updated-time-string";

    private String mLatitudeLabel;
    private String mLongitudeLabel;
    private LocationRequest mLocationRequest;
    private LocationSettingsRequest mLocationSettingsRequest;
    private Boolean mRequestingLocationUpdates;
    private SettingsClient mSettingsClient;
    private FusedLocationProviderClient mFusedLocationClient;

    private LocationCallback mLocationCallback;
    private Location mCurrentLocation;
    private String mLastUpdateTimeLabel;
    private String mLastUpdateTime;

    Button getLocation;
    CardView restaurants, monuments, oci, nocturn;
    View mView;
    TextView city_location;
    double latitude, longitude;
    MainActivity ma;

    FirebaseAuth mAuth;
    FirebaseFirestore mStore;
    String userID;
    DocumentReference documentReference;


    @Nullable
    @Override
    public View onCreateView( LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mView = inflater.inflate(R.layout.fragment_home, container, false);

        ma = (MainActivity)getActivity();

        mAuth = FirebaseAuth.getInstance();
        mStore = FirebaseFirestore.getInstance();
        userID = mAuth.getCurrentUser().getUid();
        initLocationUser();

        restaurants = mView.findViewById(R.id.cardivew_r);
        monuments = mView.findViewById(R.id.cardivew_m);
        oci = mView.findViewById(R.id.cardivew_o);
        nocturn = mView.findViewById(R.id.cardivew_n);

        restaurants.setOnClickListener(this);
        monuments.setOnClickListener(this);
        oci.setOnClickListener(this);
        nocturn.setOnClickListener(this);

        city_location = mView.findViewById(R.id.c_location_nearby);
        mLatitudeLabel = getResources().getString(R.string.latitude_label);
        mLongitudeLabel = getResources().getString(R.string.longitude_label);

        getLocation = mView.findViewById(R.id.search_location);

        getLocation.setOnClickListener(this);

        mRequestingLocationUpdates = false;
        mLastUpdateTime = "";

        updateValuesFromBundle(savedInstanceState);

        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(getActivity());
        mSettingsClient = LocationServices.getSettingsClient(getActivity());

        createLocationCallback();
        createLocationRequest();
        buildLocationSettingsRequest();

        return mView;
    }

    public void initLocationUser(){
        documentReference = mStore.collection("users").document(userID);
        documentReference.addSnapshotListener(new EventListener<DocumentSnapshot>() {
            @Override
            public void onEvent(@Nullable DocumentSnapshot snapshot, @Nullable FirebaseFirestoreException e) {
                if (documentReference != null && snapshot != null) {
                    if(!snapshot.getString("currentLocation").toString().equals("-")) {
                        city_location.setText(snapshot.get("currentLocation").toString());
                        ma.setLocation(snapshot.get("currentLocation").toString());
                        ma.setLocationActivated(true);
                    }
                }
            }
        });
    }

    private void updateValuesFromBundle(Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            if (savedInstanceState.keySet().contains(KEY_REQUESTING_LOCATION_UPDATES)) {
                mRequestingLocationUpdates = savedInstanceState.getBoolean(
                        KEY_REQUESTING_LOCATION_UPDATES);
            }

            // Update the value of mCurrentLocation from the Bundle and update the UI to show the
            // correct latitude and longitude.
            if (savedInstanceState.keySet().contains(KEY_LOCATION)) {
                // Since KEY_LOCATION was found in the Bundle, we can be sure that mCurrentLocation
                // is not null.
                mCurrentLocation = savedInstanceState.getParcelable(KEY_LOCATION);
            }

            // Update the value of mLastUpdateTime from the Bundle and update the UI.
            if (savedInstanceState.keySet().contains(KEY_LAST_UPDATED_TIME_STRING)) {
                mLastUpdateTime = savedInstanceState.getString(KEY_LAST_UPDATED_TIME_STRING);
            }
            updateLocationUI();
        }
    }


    @Override
    public void onResume() {
        super.onResume();
        // Within {@code onPause()}, we remove location updates. Here, we resume receiving
        // location updates if the user has requested them.

        if (!checkPermissions()) {
            requestPermissions();
        }
        if(ma.getLocationActivated() && !city_location.getText().equals(ma.getLocation())){
            city_location.setText(ma.getLocation());

        }

    }

    @Override
    public void onClick(View view) {
        Bundle bundle = new Bundle();
        switch (view.getId()){
            case R.id.cardivew_r:
                String Restaurants = "Restaurants";
                bundle.putString("Type_Event", Restaurants);
                goEventFragment(bundle);
                break;
            case R.id.cardivew_m:
                String Monuments = "Llocs d'Interès";
                bundle.putString("Type_Event", Monuments);
                goEventFragment(bundle);
                break;
            case R.id.cardivew_o:
                String Oci = "Oci";
                bundle.putString("Type_Event", Oci);
                goEventFragment(bundle);
                break;
            case R.id.cardivew_n:
                String Nocturn = "Nocturn";
                bundle.putString("Type_Event", Nocturn);
                goEventFragment(bundle);
                break;

            case R.id.search_location:
                if (!mRequestingLocationUpdates) {
                    mRequestingLocationUpdates = true;
                    startLocationUpdates();
                }
                ma.setLocationActivated(true);
                break;
        }
    }



    public void goEventFragment(Bundle bundle){
        Fragment selectedFragment = new EventFragment();
        selectedFragment.setArguments(bundle);
        getFragmentManager().beginTransaction().replace(R.id.fragment_main, selectedFragment).commit();
    }

    @Override
    public void onPause() {
        super.onPause();

        // Remove location updates to save battery.
        stopLocationUpdates();
    }

    private void startLocationUpdates() {
        // Begin by checking if the device has the necessary location settings.
        mSettingsClient.checkLocationSettings(mLocationSettingsRequest)
                .addOnSuccessListener(getActivity(), new OnSuccessListener<LocationSettingsResponse>() {
                    @Override
                    public void onSuccess(LocationSettingsResponse locationSettingsResponse) {
                        Log.i(TAG, "All location settings are satisfied.");

                        //noinspection MissingPermission
                        mFusedLocationClient.requestLocationUpdates(mLocationRequest,
                                mLocationCallback, Looper.myLooper());

                        updateLocationUI();
                    }
                })
                .addOnFailureListener(getActivity(), new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        int statusCode = ((ApiException) e).getStatusCode();
                        switch (statusCode) {
                            case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:
                                Log.i(TAG, "Location settings are not satisfied. Attempting to upgrade " +
                                        "location settings ");
                                try {
                                    // Show the dialog by calling startResolutionForResult(), and check the
                                    // result in onActivityResult().
                                    ResolvableApiException rae = (ResolvableApiException) e;
                                    rae.startResolutionForResult(getActivity(), REQUEST_CHECK_SETTINGS);
                                } catch (IntentSender.SendIntentException sie) {
                                    Log.i(TAG, "PendingIntent unable to execute request.");
                                }
                                break;
                            case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
                                String errorMessage = "Location settings are inadequate, and cannot be " +
                                        "fixed here. Fix in Settings.";
                                Log.e(TAG, errorMessage);
                                Toast.makeText(getActivity(), errorMessage, Toast.LENGTH_LONG).show();
                                mRequestingLocationUpdates = false;
                        }

                        updateLocationUI();
                    }
                });
    }

    private void createLocationCallback() {
        mLocationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                super.onLocationResult(locationResult);

                mCurrentLocation = locationResult.getLastLocation();
                mLastUpdateTime = DateFormat.getTimeInstance().format(new Date());
                updateLocationUI();
            }
        };
    }

    private void updateLocationUI(){
        if (mCurrentLocation != null) {
            Geocoder geocoder = new Geocoder (getContext(), Locale.getDefault());
            try{
                List<Address> addresses = geocoder.getFromLocation(mCurrentLocation.getLatitude(), mCurrentLocation.getLongitude(), 1);
                Address obj = addresses.get(0);
                String add = obj.getLocality();
                takeOfAccents(add);
            }catch (IOException e){
                e.printStackTrace();
                Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
            }

        }
    }

    private void createLocationRequest() {
        mLocationRequest = new LocationRequest();
        mLocationRequest.setInterval(UPDATE_INTERVAL_IN_MILLISECONDS);
        mLocationRequest.setFastestInterval(FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
    }

    private void buildLocationSettingsRequest(){
        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder();
        builder.addLocationRequest(mLocationRequest);
        mLocationSettingsRequest = builder.build();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            // Check for the integer request code originally supplied to startResolutionForResult().
            case REQUEST_CHECK_SETTINGS:
                switch (resultCode) {
                    case Activity.RESULT_OK:
                        Log.i(TAG, "User agreed to make required location settings changes.");
                        // Nothing to do. startLocationupdates() gets called in onResume again.
                        break;
                    case Activity.RESULT_CANCELED:
                        Log.i(TAG, "User chose not to make required location settings changes.");
                        mRequestingLocationUpdates = false;
                        updateLocationUI();
                        break;
                }
                break;
        }
    }

    private void requestPermissions() {
        String[] permissions = {Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION};
        if(ContextCompat.checkSelfPermission(getActivity().getApplicationContext(),
                FINE_LOCATION) == PackageManager.PERMISSION_GRANTED){
            if(ContextCompat.checkSelfPermission(getActivity().getApplicationContext(),
                    COURSE_LOCATION) == PackageManager.PERMISSION_GRANTED){

            }else{
                ActivityCompat.requestPermissions(getActivity(),
                        permissions,
                        LOCATION_PERMISSION_REQUEST_CODE);
            }
        }else{
            ActivityCompat.requestPermissions(getActivity(),
                    permissions,
                    LOCATION_PERMISSION_REQUEST_CODE);
        }
    }

    private boolean checkPermissions() {
        int permissionState = ActivityCompat.checkSelfPermission(getActivity(),
                Manifest.permission.ACCESS_FINE_LOCATION);
        return permissionState == PackageManager.PERMISSION_GRANTED;
    }

    private void stopLocationUpdates() {
        if (!mRequestingLocationUpdates) {
            Log.d(TAG, "stopLocationUpdates: updates never requested, no-op.");
            return;
        }

        // It is a good practice to remove location requests when the activity is in a paused or
        // stopped state. Doing so helps battery performance and is especially
        // recommended in applications that request frequent location updates.
        mFusedLocationClient.removeLocationUpdates(mLocationCallback)
                .addOnCompleteListener(getActivity(), new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        mRequestingLocationUpdates = false;
                    }
                });
    }

    public void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putBoolean(KEY_REQUESTING_LOCATION_UPDATES, mRequestingLocationUpdates);
        savedInstanceState.putParcelable(KEY_LOCATION, mCurrentLocation);
        savedInstanceState.putString(KEY_LAST_UPDATED_TIME_STRING, mLastUpdateTime);
        super.onSaveInstanceState(savedInstanceState);
    }

    public void takeOfAccents(String text){
        String string = Normalizer.normalize(text, Normalizer.Form.NFD);
        string = string.replaceAll("[^\\p{ASCII}]", "");
        if(!string.equals(ma.getLocation())){
            city_location.setText(string);
            ma.setLocation(string);
            actualizarInfoUser();
            ma.sendOnMainChannel(mView);
        }
        stopLocationUpdates();
    }

    public void actualizarInfoUser(){
        //First we have to subscribe and unsubscribe from Topics
        if(!city_location.getText().toString().equals("-")) {
            FirebaseMessaging.getInstance().unsubscribeFromTopic(city_location.getText().toString());
        }
        checkUserInfo();

        String currentLocation = ma.getLocation();
        Map<String, Object> user = new HashMap<>();
        user.put("currentLocation", currentLocation);
        documentReference.set(user, SetOptions.merge());
        FirebaseMessaging.getInstance().subscribeToTopic(ma.getLocation());

        checkUserInfo();
    }

    public void checkUserInfo(){
        final DocumentReference document = mStore.collection("users").document(userID);
        Log.d("HELLO", "HELLO DARLING!");

        document.addSnapshotListener(new EventListener<DocumentSnapshot>() {
            @Override
            public void onEvent(@Nullable DocumentSnapshot value, @Nullable FirebaseFirestoreException error) {
                Log.d("USER-NAME", value.getString("name") +" "+ value.getString("surname"));
                Log.d("USER-CURRENT_LOCATION", value.getString("currentLocation"));

            }
        });
    }
}

File: app/src/main/java/com/example/whattodo/HomeFragment.kt
package com.example.whattodo

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.content.IntentSender.SendIntentException
import android.content.pm.PackageManager
import android.location.Geocoder
import android.location.Location
import android.os.Bundle
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.cardview.widget.CardView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.*
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.google.firebase.messaging.FirebaseMessaging
import java.io.IOException
import java.text.DateFormat
import java.text.Normalizer
import java.util.*

class HomeFragment : Fragment(), View.OnClickListener {

    private lateinit var mLatitudeLabel: String
    private lateinit var mLongitudeLabel: String
    private lateinit var mLocationRequest: LocationRequest
    private lateinit var mLocationSettingsRequest: LocationSettingsRequest

    private var mRequestingLocationUpdates: Boolean? = null
    private lateinit var mSettingsClient: SettingsClient
    private lateinit var mFusedLocationClient: FusedLocationProviderClient
    private lateinit var mLocationCallback: LocationCallback
    private lateinit var mCurrentLocation: Location
    private val mLastUpdateTimeLabel: String? = null
    private lateinit var mLastUpdateTime: String

    private lateinit var getLocation: Button
    private lateinit var restaurants: CardView
    private lateinit var monuments: CardView
    private lateinit var oci: CardView
    private lateinit var nocturn: CardView

    private lateinit var viewOfLayout: View

    private lateinit var city_location: TextView
    var latitude = 0.0
    var longitude = 0.0

    private lateinit var ma: MainActivity
    private lateinit var mAuth: FirebaseAuth
    private lateinit var mStore: FirebaseFirestore

    private lateinit var userID: String
    private lateinit var documentReference: DocumentReference

    override fun onCreateView(inflater: LayoutInflater,container: ViewGroup?,savedInstanceState: Bundle? ): View? {
        viewOfLayout = inflater.inflate(R.layout.fragment_home, container, false)
        ma = (activity as MainActivity)
        mAuth = FirebaseAuth.getInstance()
        mStore = FirebaseFirestore.getInstance()
        userID = mAuth!!.currentUser!!.uid
        initLocationUser()

        restaurants = viewOfLayout.findViewById(R.id.cardivew_r)
        monuments = viewOfLayout.findViewById(R.id.cardivew_m)
        oci = viewOfLayout.findViewById(R.id.cardivew_o)
        nocturn = viewOfLayout.findViewById(R.id.cardivew_n)

        restaurants.setOnClickListener(this)
        monuments.setOnClickListener(this)
        oci.setOnClickListener(this)
        nocturn.setOnClickListener(this)

        city_location = viewOfLayout.findViewById(R.id.c_location_nearby)
        mLatitudeLabel = resources.getString(R.string.latitude_label)
        mLongitudeLabel = resources.getString(R.string.longitude_label)

        getLocation = viewOfLayout.findViewById(R.id.search_location)
        getLocation.setOnClickListener(this)

        mRequestingLocationUpdates = false
        mLastUpdateTime = ""
        updateValuesFromBundle(savedInstanceState)

        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(requireActivity())
        mSettingsClient = LocationServices.getSettingsClient(requireActivity())

        createLocationCallback()
        createLocationRequest()
        buildLocationSettingsRequest()

        return viewOfLayout
    }

    fun initLocationUser() {
        documentReference = mStore!!.collection("users").document(userID!!)
        documentReference!!.addSnapshotListener { snapshot, e ->
            if (documentReference != null && snapshot != null) {
                if (snapshot.getString("currentLocation").toString() != "-") {
                    city_location!!.text = snapshot["currentLocation"].toString()
                    ma!!.location = snapshot["currentLocation"].toString()
                    ma!!.locationActivated = true
                }
            }
        }
    }

    private fun updateValuesFromBundle(savedInstanceState: Bundle?) {
        if (savedInstanceState != null) {
            if (savedInstanceState.keySet().contains(KEY_REQUESTING_LOCATION_UPDATES)) {
                mRequestingLocationUpdates = savedInstanceState.getBoolean(
                    KEY_REQUESTING_LOCATION_UPDATES
                )
            }

            // Update the value of mCurrentLocation from the Bundle and update the UI to show the
            // correct latitude and longitude.
            if (savedInstanceState.keySet().contains(KEY_LOCATION)) {
                // Since KEY_LOCATION was found in the Bundle, we can be sure that mCurrentLocation
                // is not null.
                mCurrentLocation = savedInstanceState.getParcelable(KEY_LOCATION)!!
            }

            // Update the value of mLastUpdateTime from the Bundle and update the UI.
            if (savedInstanceState.keySet().contains(KEY_LAST_UPDATED_TIME_STRING)) {
                mLastUpdateTime = savedInstanceState.getString(KEY_LAST_UPDATED_TIME_STRING).toString()
            }
            updateLocationUI()
        }
    }

    override fun onResume() {
        super.onResume()
        // Within {@code onPause()}, we remove location updates. Here, we resume receiving
        // location updates if the user has requested them.
        if (!checkPermissions()) {
            requestPermissions()
        }
        if (ma!!.locationActivated && city_location!!.text != ma!!.location) {
            city_location!!.text = ma!!.location
        }
    }

    override fun onClick(view: View) {
        val bundle = Bundle()
        when (view.id) {
            R.id.cardivew_r -> {
                val Restaurants = "Restaurants"
                bundle.putString("Type_Event", Restaurants)
                goEventFragment(bundle)
            }
            R.id.cardivew_m -> {
                val Monuments = "Llocs d'Interès"
                bundle.putString("Type_Event", Monuments)
                goEventFragment(bundle)
            }
            R.id.cardivew_o -> {
                val Oci = "Oci"
                bundle.putString("Type_Event", Oci)
                goEventFragment(bundle)
            }
            R.id.cardivew_n -> {
                val Nocturn = "Nocturn"
                bundle.putString("Type_Event", Nocturn)
                goEventFragment(bundle)
            }
            R.id.search_location -> {
                if (!mRequestingLocationUpdates!!) {
                    mRequestingLocationUpdates = true
                    startLocationUpdates()
                }
                ma!!.locationActivated = true
            }
        }
    }

    fun goEventFragment(bundle: Bundle?) {
        val selectedFragment: Fragment = EventFragment()
        selectedFragment.arguments = bundle
        requireFragmentManager().beginTransaction().replace(R.id.fragment_main, selectedFragment).commit()
    }

    override fun onPause() {
        super.onPause()

        // Remove location updates to save battery.
        stopLocationUpdates()
    }

    @SuppressLint("MissingPermission")
    private fun startLocationUpdates() {
        // Begin by checking if the device has the necessary location settings.
        mSettingsClient!!.checkLocationSettings(mLocationSettingsRequest)
            .addOnSuccessListener(requireActivity()) {
                Log.i(TAG, "All location settings are satisfied.")
                mFusedLocationClient!!.requestLocationUpdates(mLocationRequest,mLocationCallback, Looper.myLooper())
            }
            .addOnFailureListener(requireActivity()) { e ->
                val statusCode = (e as ApiException).statusCode
                when (statusCode) {
                    LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {
                        Log.i(
                            TAG, "Location settings are not satisfied. Attempting to upgrade " +
                                    "location settings "
                        )
                        try {
                            // Show the dialog by calling startResolutionForResult(), and check the
                            // result in onActivityResult().
                            val rae = e as ResolvableApiException
                            rae.startResolutionForResult(activity, REQUEST_CHECK_SETTINGS)
                        } catch (sie: SendIntentException) {
                            Log.i(TAG, "PendingIntent unable to execute request.")
                        }
                    }
                    LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> {
                        val errorMessage = "Location settings are inadequate, and cannot be " +
                                "fixed here. Fix in Settings."
                        Log.e(TAG, errorMessage)
                        Toast.makeText(activity, errorMessage, Toast.LENGTH_LONG).show()
                        mRequestingLocationUpdates = false
                    }
                }
                updateLocationUI()
            }
    }

    private fun createLocationCallback() {
        mLocationCallback = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult) {
                super.onLocationResult(locationResult)
                Toast.makeText(context,"LOCATION-RESULT: " + locationResult.lastLocation.toString(), Toast.LENGTH_SHORT).show()
                mCurrentLocation = locationResult.lastLocation
                mLastUpdateTime = DateFormat.getTimeInstance().format(Date())
                updateLocationUI()
            }
        }
    }

    private fun updateLocationUI() {
        if (mCurrentLocation != null) {
            val geocoder = Geocoder(context, Locale.getDefault())
            try {
                val addresses = geocoder.getFromLocation(
                    mCurrentLocation!!.latitude, mCurrentLocation!!.longitude, 1
                )
                val obj = addresses[0]
                val add = obj.locality
                takeOfAccents(add)
            } catch (e: IOException) {
                e.printStackTrace()
                Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun createLocationRequest() {
        mLocationRequest = LocationRequest()
        mLocationRequest!!.interval = UPDATE_INTERVAL_IN_MILLISECONDS
        mLocationRequest!!.fastestInterval = FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS
        mLocationRequest!!.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
    }

    private fun buildLocationSettingsRequest() {
        val builder = LocationSettingsRequest.Builder()
        builder.addLocationRequest(mLocationRequest!!)
        mLocationSettingsRequest = builder.build()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_CHECK_SETTINGS -> when (resultCode) {
                Activity.RESULT_OK -> Log.i(
                    TAG,
                    "User agreed to make required location settings changes."
                )
                Activity.RESULT_CANCELED -> {
                    Log.i(TAG, "User chose not to make required location settings changes.")
                    mRequestingLocationUpdates = false
                    updateLocationUI()
                }
            }
        }
    }

    private fun requestPermissions() {
        val permissions = arrayOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        )
        if (ContextCompat.checkSelfPermission(
                requireActivity().applicationContext,
                FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            if (ContextCompat.checkSelfPermission(
                    requireActivity().applicationContext,
                    COURSE_LOCATION
                ) == PackageManager.PERMISSION_GRANTED
            ) {
            } else {
                ActivityCompat.requestPermissions(
                    requireActivity(),
                    permissions,
                    LOCATION_PERMISSION_REQUEST_CODE
                )
            }
        } else {
            ActivityCompat.requestPermissions(
                requireActivity(),
                permissions,
                LOCATION_PERMISSION_REQUEST_CODE
            )
        }
    }

    private fun checkPermissions(): Boolean {
        val permissionState = ActivityCompat.checkSelfPermission(
            requireActivity(),
            Manifest.permission.ACCESS_FINE_LOCATION
        )
        return permissionState == PackageManager.PERMISSION_GRANTED
    }

    private fun stopLocationUpdates() {
        if (!mRequestingLocationUpdates!!) {
            Log.d(TAG, "stopLocationUpdates: updates never requested, no-op.")
            return
        }

        // It is a good practice to remove location requests when the activity is in a paused or
        // stopped state. Doing so helps battery performance and is especially
        // recommended in applications that request frequent location updates.
        mFusedLocationClient!!.removeLocationUpdates(mLocationCallback)
            .addOnCompleteListener(requireActivity()) { mRequestingLocationUpdates = false }
    }

    /*
    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putBoolean(KEY_REQUESTING_LOCATION_UPDATES, mRequestingLocationUpdates!!)
        savedInstanceState.putParcelable(KEY_LOCATION, mCurrentLocation)
        savedInstanceState.putString(KEY_LAST_UPDATED_TIME_STRING, mLastUpdateTime)
        super.onSaveInstanceState(savedInstanceState)
    }
    */

    fun takeOfAccents(text: String?) {
        var string = Normalizer.normalize(text, Normalizer.Form.NFD)
        string = string.replace("[^\\p{ASCII}]".toRegex(), "")
        if (string != ma!!.location) {
            city_location!!.text = string
            ma!!.location = string
            actualizarInfoUser()
            ma!!.sendOnMainChannel(viewOfLayout)
        }
        stopLocationUpdates()
    }

    fun actualizarInfoUser() {
        //First we have to subscribe and unsubscribe from Topics
        if (city_location!!.text.toString() != "-") {
            FirebaseMessaging.getInstance().unsubscribeFromTopic(city_location!!.text.toString())
        }
        checkUserInfo()
        val currentLocation = ma!!.location
        val user: MutableMap<String, Any?> = HashMap()
        user["currentLocation"] = currentLocation
        documentReference!![user] = SetOptions.merge()
        FirebaseMessaging.getInstance().subscribeToTopic(ma!!.location!!)
        checkUserInfo()
    }

    fun checkUserInfo() {
        val document = mStore!!.collection("users").document(
            userID!!
        )
        document.addSnapshotListener { value, error ->
            Log.d("USER-NAME", value!!.getString("name") + " " + value.getString("surname"))
            Log.d("USER-CURRENT_LOCATION", value.getString("currentLocation")!!)
        }
    }

    companion object {
        private const val UPDATE_INTERVAL_IN_MILLISECONDS: Long = 10000
        private const val FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS =
            UPDATE_INTERVAL_IN_MILLISECONDS / 2
        private val TAG = MainActivity::class.java.simpleName
        private const val REQUEST_CHECK_SETTINGS = 0x1
        private const val LOCATION_PERMISSION_REQUEST_CODE = 34
        private const val FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION
        private const val COURSE_LOCATION = Manifest.permission.ACCESS_COARSE_LOCATION
        private const val KEY_REQUESTING_LOCATION_UPDATES = "requesting-location-updates"
        private const val KEY_LOCATION = "location"
        private const val KEY_LAST_UPDATED_TIME_STRING = "last-updated-time-string"
    }
}